src/gui/torrentcreatordialog.cpp

```diff
@@ -233,8 +229,14 @@ void TorrentCreatorDialog::onCreateButtonClicked()
         , m_ui->URLSeedsList->toPlainText().split(u'\n', Qt::SkipEmptyParts)
     };
 
-    // run the creator thread
-    m_creatorThread->create(params);
+    auto *torrentCreator = new BitTorrent::TorrentCreator(params);
+    connect(this, &QDialog::finished, torrentCreator, &BitTorrent::TorrentCreator::requestInterruption);
```

Comment:
It works the same way as before, just not hidden in the depths of QThread primitives. Before, ~QThread was setting a similar flag in the destructor(https://doc.qt.io/qt-6/qthread.html#dtor.QThread) and TorrentCreatorThread gracefully aborted the current task. Now, a similar flag is set via a signal: so when `m_interruptionRequested` is set, the run() function will abort the operation after hashing the next piece.

Reply:
I've read a bit of documentation and I'm a bit puzzled. I think that both the change introduce by this commit, but also the current upstream version which uses QThreads are buggy. In both cases the thread affinity of the QObject (QThread in the  upstream version; the TorrentCreator in the new version) is in the thread of the dialog. Meaning that all `emit()` signals will use `Qt::DirectConnection` because the signal is seen by QT as remaining within the same thread. I think that the race condition is just so difficult to trigger in real-world use-cases by users, that nobody realized about it. I wonder if setting the connection type explicitly to `Qt::QueuedConnection` is enough to solve the issue.

Reply:
`QThread::requestInterruption()` is thread-safe according to documentation.

Reply:
> Meaning that all `emit()` signals will use `Qt::DirectConnection` because the signal is seen by QT as remaining within the same thread.

Apparently, you didn't quite get it right. Qt compares the current thread (where signal is emitted from) and the thread of the context (i.e. the receiver). So the signals emitted by the torrent creator from the `run()` method (when it is executed in another thread) are processed by the Dialog as queued connections in Dialog threaf.
The only problem is this one connection.

Reply:
Thank you for explaining. According to my understanding, this should be OK now. This particular connection will be executed as direct. So the boolean will be set from within the dialog thread (which is also the thread who will delete the QRunnable within the context of QThreadPool). While it's a different thread than the one where creation happens, I don't think it's an unsafe operation, as I described in the discussion on another review comment. But I can do it explicitly strongly thread-safe if you consider it's better this way. Please correct me if I'm still miss-understanding something.